Add ICUCollationFieldMapper

mattweber opened the pull request:
Adds a new "icu_collation" field type that exposes Lucene's
ICUCollationDocValuesField.  ICUCollationDocValuesField is the replacement
for ICUCollationKeyFilter which has been deprecated since Lucene 5.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

mattweber commented:
Closes #13148 and would allow #15827 to proceed.

/cc @dadoonet @clintongormley

javanna commented:
@jpountz can you review this or delegate further?

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationFieldMapper.java:
```diff
@@ -0,0 +1,623 @@
+            @Override
+            public double parseDouble(String value, boolean roundUp, LongSupplier now) {
+                return Double.parseDouble(value);
+            }
```
parseLong and parseDouble should just throw an exception

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationFieldMapper.java:
```diff
@@ -0,0 +1,623 @@
+            @Override
+            public String format(double value) {
+                return Double.toString(value);
+            }
```
format on long and double values should just throw an exception

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationFieldMapper.java:
```diff
@@ -0,0 +1,623 @@
+    private final Collator collator;
+
+    // each thread needs its own ICUCollationDocValuesField because the internally cloned collator is not thread-safe
+    private final ThreadLocal<ICUCollationDocValuesField> perThreadICUCollationField;
```
we avoid using thread locals in other mappers since it would break if you pass an array of values as all values would share the same field instance. I think we should just create field instances on demand.

mattweber replied:
I did this because of [this comment](https://github.com/apache/lucene-solr/blob/master/lucene/analysis/icu/src/java/org/apache/lucene/collation/ICUCollationDocValuesField.java#L45).  Not sure how expensive cloning a `Collator` is but it will happen for every value being indexed if we create them on demand right?

jpountz replied:
ICUCollationDocValuesField is just a convenience class, you could do the same by calling the collator manually and then pass the result to a SortedDocValuesField. Maybe this is what we should do? This would also be helpful to avoid computing the collation twice if we want to index the value as well.

mattweber replied:
Sounds good, I will give that try.  Thanks!

mattweber commented:
Thanks for the review @jpountz!  I will work on your requested changes and get back to you.  I think adding index support is a great idea, I had totally missed `ICUCollationKeyAnalyzer`.

mattweber commented:
@jpountz Ready for another review.  Still working on the docs.

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,738 @@
+
+            if (collator != null) {
+                RawCollationKey key;
+                synchronized (collatorLock) {
```
I see this in the Collator javadocs: `A frozen Collator is immutable and thread-safe.` Should we freeze the collator and remove the lock?

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,738 @@
+                int decodedLength = IndexableBinaryStringTools.getDecodedLength(encoded, 0, encoded.length);
+                byte[] decoded = new byte[decodedLength];
+                IndexableBinaryStringTools.decode(encoded, 0, encoded.length, decoded, 0, decodedLength);
+                return new BytesRef(decoded);
```
Maybe format and parseBytesRef should reuse the collator? Otherwise the keys you would get eg. in aggregations would make little sense. This requires additional serialization logic to serialize the configuration of the collator, you can look at `DocValueFormat.DateTime` to see how to make it work.

mattweber replied:
For `format`, isn't that coming up with a string representation of the indexed value?  The indexed values are not valid utf-8 in some cases thus the reason I had to use the same approach the old token filter used.  Not sure how I would utilize the `Collator` here.  On that note, shouldn't `parseBytesRef` be parsing what `format` generated?

jpountz replied:
Oh sorry you are right, there is no way we can get the original string back. The approach you have been folowing is right.

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,738 @@
+                        builder.alternate(XContentMapValues.nodeStringValue(fieldNode, null));
+                        iterator.remove();
+                        break;
+                    case "caseLevel":
```
can you use underscore case instead? We are progressively moving out from camel case in APIs.

jpountz replied:
same for caseFirst, variableTop and hiraganaQuaternaryMode

mattweber commented:
@jpountz pushed another iteration.  Only thing missing here is using the `Collator` in the `DocValueFormat` which I am not sure will work, please see comments above.  Also, not sure if I added testCompile dependency on core tests correctly, please verify.

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,755 @@
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            throw new UnsupportedOperationException();
```
maybe this one should be supported?

mattweber commented:
@jpountz I enabled range query support, moved the `FieldTypeTestCase` into the `test-framework` and added docs.

I rebased with latest changes from master since we were getting a little behind.  I did not squash commits yet, the last three are the new ones.

jpountz commented on core/build.gradle:
```diff
@@ -99,6 +99,14 @@ dependencies {
+dependencies {
+  testOutput sourceSets.test.output
+}
+
```
this should not be necessary anymore?

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,750 @@
+        this.numeric = numeric;
+        this.variableTop = variableTop;
+        this.hiraganaQuaternaryMode = hiraganaQuaternaryMode;
+        this.collator = collator.isFrozen() ? collator : collator.freeze();
```
maybe we should just assert that the collator is frozen, there don't seem to be code paths that provide mutable collators?

jpountz commented on plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java:
```diff
@@ -0,0 +1,750 @@
+    @Override
+    protected ICUCollationKeywordFieldMapper clone() {
+        return (ICUCollationKeywordFieldMapper) super.clone();
+    }
```
do we really need to override it?

mattweber commented:
@jpountz Done.  Let me know and I will rebase + squash.

jpountz commented:
@mattweber Thanks! I don't think I need you to rebase+squash, I'll let Github take care of doing that for us.

@elasticmachine please test

jpountz commented:
@elasticmachine test this please.

mattweber commented:
@jpountz pushed line length fix for failed test

jpountz commented:
@elasticmachine test this please.

jpountz commented:
Merged to master and 5.x. Thanks @mattweber !
